Implement scoring for SR2025

PeterJCLaw opened the pull request:
This introduces the scoring & conversion logic and the input UI.

CI failure is unrelated -- caused by lack of league matches in the schedule.

Compare to https://github.com/srobo/arena/pull/6

Contributes to https://github.com/srobo/tasks/issues/1402
Relates to https://github.com/srobo/tasks/issues/1439

Fixes https://github.com/srobo/tasks/issues/1401

![image](https://github.com/user-attachments/assets/5c6ca438-9a0d-4d07-b3aa-0598303cad94)

PeterJCLaw committed:
Initial scoring implementation

This adds the basic data conversion logic for the form, though
doesn't add a UI template yet, as well as a first pass at the scoring
logic itself, though without any validation.

PeterJCLaw committed:
Add initial input validation

PeterJCLaw committed:
Whitespace

PeterJCLaw committed:
Catch the right exception

PeterJCLaw committed:
Improve these error messages

PeterJCLaw committed:
Add tests and fix a silly bug they highlighted

PeterJCLaw committed:
Add overlooked movement point handling

PeterJCLaw committed:
First pass at a scorer UI

Completely untested.

PeterJCLaw committed:
Connect up input-type specific validation

This relies on v1.6 of the scorer.

PeterJCLaw committed:
Improve spacing of district inputs

This makes them more uniform and (combined with v1.7 of the scorer)
helps ensure the input placeholders are visible.

samjmartin04 commented:
Looks good. Zones need to be rotated around by one anticlockwise so that green zone 0 is bottom left to match the physical scoresheet

PeterJCLaw committed:
Rotate the corners to match this year's arrangement

PeterJCLaw commented:
> Looks good. Zones need to be rotated around by one anticlockwise so that green zone 0 is bottom left to match the physical scoresheet

a9ff512d0f52d75b442d9ac77299d3be01a4b6e3

PeterJCLaw commented:
> Looks good. Zones need to be rotated around by one anticlockwise so that green zone 0 is bottom left to match the physical scoresheet

Good spot! a9ff512d0f52d75b442d9ac77299d3be01a4b6e3

samjmartin04 commented on scoring/score.py:
```diff
@@ -13,21 +24,158 @@ def __init__(self, message: str, *, code: str) -> None:
+        num_tokens = district['pallet_counts'][colour]
+        score = num_tokens * DISTRICT_SCORE_MAP[name]
+
+        if colour in district['highest']:
+            # Points are doubled for the team owning the highest pallet
+            score *= 2
```
Is `district['highest']` restricted to only one zone letter at the point of data entry? Either way, I wonder whether we should check to see if there's only 1 team that is highest before rewarding this? I could see a marshal entering e.g. 'GY' for highest when they are both highest.

PeterJCLaw replied:
There _should_ be a validation check for that already. If any of the validation checks fail then computing the scores is abandoned.

PeterJCLaw replied:
This is validated in the UI for fast feedback (though by design that's non-blocking) and in the checks in the scorer (specifically https://github.com/srobo/sr2025-comp/pull/1/files#diff-b89188034b113b9776c2ad5f796fc9c0621c4dd01fae56f243ed3ce3b7d46f30R101-R114).

samjmartin04 replied:
Okay, regarding the second half of my comment then - do we want those checks? I can see it as being quite valid for a marshal to enter two letters for two pallets that are both stacked on another pallet (despite the 25 mm higher clause) and I think the score entry being able to treat that case as no highest token probably would be useful?

PeterJCLaw replied:
Historically we have viewed the scoring logic as applying the _scoring_ rules only, not making judgements about the state of the arena. The checks are intended to catch cases which are incompatible with the rules, rather than to provide an opinion about the state.

From a technical standpoint it's feasible to change that, though I'm not sure it's a good idea. I can see the (IMO small) benefit of recording further information about the state of the arena though.

I think my concern around changing that is that it feels like it could be really easy for the wrong thing to be noted down if there's any lack of clarity around where the judgement-calls will be being applied. I don't have a good feel for how likely that is.

samjmartin04 replied:
Decision - we only want marshals to list a highest token if a *single* token is highest, so only having 1 pallet allowed to be highest is desired

samjmartin04 commented on scoring/score.py:
```diff
@@ -13,21 +24,158 @@ def __init__(self, message: str, *, code: str) -> None:
+                code='impossible_highest_pallet',
+            )
+
+        missing_highest = {}
+        for name, district in self._districts.items():
+            highest = district['highest']
+            pallet_counts = district['pallet_counts']
+            if len(pallet_counts) == 1 and not highest:
+                pallet, = pallet_counts.keys()
+                missing_highest[name] = pallet
+        if missing_highest:
+            raise InvalidScoresheetException(
+                f"Some districts with pallets from a single team are missing "
+                "specification of the highest.\n"
+                f"{missing_highest!r}",
+                code='missing_highest_pallet',
+            )
```
Is this checking that if there's at least one pallet in the scoring zone that there should be one listed in highest? Because there are scenarios where there is only one pallet in the scoring zone and it does not count as being highest (e.g. if it's on the floor)

PeterJCLaw replied:
This is checking the case where there's pallets from _exactly one_ team present in the scoring zone, in which case (I believe?) then that team's pallets _must_ be the highest.

(explanatory comments added in https://github.com/srobo/sr2025-comp/pull/1/commits/ae266736bf2966b80403da01651eb2bd3a0a3a65)

samjmartin04 replied:
Ah okay. I think in the case of only one pallet on the floor, no pallet is highest as pallets must not be touching the floor to count as highest, so no pallets would be doubled in that scenario

PeterJCLaw replied:
:facepalm: d'oh, forgot about that rule. Yeah, we can remove this check then and I thiiink everything else still holds?

samjmartin04 replied:
Yeah I think we can remove it. There's a consideration here for how marshals will actually end up doing the highest - I can see an argument for telling them to write a single line to indicate no highest to show that they have checked, but I suppose score entry can just treat these as a blank square when actually entering them

PeterJCLaw replied:
0cc5b57c2483655ec678f53395fa85d460b7f379 removes that check. 4ca07e188f0fe0f9a9149fe5975e2f479eec5a6b introduces the inverse check, only for the outer districts.

samjmartin04 commented on scoring/score.py:
```diff
@@ -13,21 +24,158 @@ def __init__(self, message: str, *, code: str) -> None:
+            totals += district['pallet_counts']
+        if any(x > TOKENS_PER_ZONE for x in totals.values()):
+            raise InvalidScoresheetException(
+                f"Too many pallets of some kinds specified, must be no more "
+                f"than {TOKENS_PER_ZONE} of each type.\n"
+                f"{totals!r}",
```
Could we say **which** pallet has been listed too many times here?

PeterJCLaw replied:
I'd hoped that that would be sufficiently obvious from the totals. The error message here looks like:
> Too many pallets of some kinds specified, must be no more than 6 of each type.
> {'P': 7, 'G': 3}

I guess we could say like:
> Too many 'P' pallets specified, must be no more than 6 of each type.
> Totals: {'P': 7, 'G': 3}

Is that what you had in mind?

PeterJCLaw replied:
9c0b0b2 does that.

samjmartin04 replied:
I like this change

PeterJCLaw committed:
Explain each of the implemented checks

This makes it easier to understand the intent of these code blocks
as well as enabling non-Python-fluent readers to consider which
checks have been implemented.

PeterJCLaw committed:
Document this helper

PeterJCLaw committed:
Include slightly more detail in this validation message

samjmartin04 commented on scoring/score.py:
```diff
@@ -13,21 +24,182 @@ def __init__(self, message: str, *, code: str) -> None:
 
     def validate(self, other_data):
-        warnings.warn("Scoresheet validation not implemented")
+        # Check that the right districts are specified.
+        if self._districts.keys() != DISTRICT_SCORE_MAP.keys():
+            missing = DISTRICT_SCORE_MAP.keys() - self._districts.keys()
+            extra = self._districts.keys() - DISTRICT_SCORE_MAP.keys()
+            detail = "Wrong districts specified."
+            if missing:
+                detail += f" Missing: {join_and(missing)}."
+            if extra:
+                detail += f" Extra: {join_and(repr(x) for x in extra)}."
+            raise InvalidScoresheetException(
+                detail,
+                code='invalid_districts',
+            )
```
What does this test do? Where are the districts specified?

PeterJCLaw replied:
This checks that the districts actually present are the expected ones. The districts are specified by `DISTRICT_SCORE_MAP` which is defined within the software (specifically it's imported from `sr2025.py`).

PeterJCLaw committed:
Pallets on the floor don't qualify to be the highest

PeterJCLaw committed:
Add check that pallets which must be on the floor are not the highest

Pallets in districts which don't have high-rises must be on the
floor if there is only one pallet in the district. This is an
imperect check since it cannot validate districts which do have
high-rises. It's probably still useful though.

samjmartin04 commented:
I'm happy with the checks we have now. Just needs a code review (from not me) to ensure that the checks are doing what we wish them to

PeterJCLaw committed:
Swap to 2×2 grid of pallet counts

Following user testing at yesterday's Tech Day this turns out to
be the preferred input format for the paper forms.

This commit fully moves over to that style -- UI, converter, storage
and scoring logic.

PeterJCLaw committed:
Add missing typing conversion

Adimote commented on scoring/converter.py:
```diff
@@ -10,12 +10,137 @@
+        """
+        Prepare a team's scoring data for saving in a score dict.
+
+        This is given a zone as form data is all keyed by zone.
```
my brain struggled to parse this line, maybe add a comma or two?

PeterJCLaw replied:
This is actually copied from the base class, but I agree could be better. How about efc9898? If we like that I'll upstream it too.

Adimote replied:
LGTM

Adimote commented on scoring/update.html:
```diff
@@ -1,5 +1,80 @@
+    </foreignObject>
+{% endmacro %}
+
+{% macro input_pallets(x, y, name, colour_symbol) %}
```
Can we maybe colour the letters, too? My mental process for "P" -> Pink, "O" -> Orange, etc.. is a bit slower than just colour matching.

PeterJCLaw replied:
Do you mean in the `input` fields in the UI? So having the letters the user is typing being in the relevant colours?

If so, I'm not sure that that would actually help -- the score-sheets which are the reference input will be in either black or blue pen and the user _does not need to know_ what the letters stand for.

Adimote replied:
Here's my proposal, add a glow to the letters, similar to the zone numbers
![image](https://github.com/user-attachments/assets/35df18b9-96ce-4fda-9a94-dedcb9b3b916)

PeterJCLaw replied:
Oh, I see. (I had also forgotten that we'd made that layout change :facepalm:). Yeah, I can definitely see this being useful in the virtual competition where we're going straight in rather than copying from the paper sheets (which as mentioned are greyscale).

I'll have a look at how easy this is to do.

PeterJCLaw replied:
Done in https://github.com/srobo/sr2025-comp/pull/1/commits/f473854eff8e228867f7b24a28ab204fa949a064

![image](https://github.com/user-attachments/assets/7bcb37b0-6779-4031-a393-3ee7a812408d)

Adimote commented on scoring/converter.py:
```diff
@@ -10,12 +10,137 @@
+                form.get(f'left_starting_zone_{zone_id}', None) is not None,
+        }
+
+    def form_district_to_score(self, form: InputForm, name: str) -> RawDistrict:
```
[nit] It doesn't look like this returns a score, it returns the raw district data, maybe we should rename the function for clarity?

PeterJCLaw replied:
Hrm, the returned object is not itself a whole score, however it is in the score-flavoured structure (rather than the form-flavoured structure passed in). Open to better ideas if we have any, though worth bearing in mind that this is following existing patterns for naming of these functions.

Adimote replied:
Feel free to submit as-is,

maybe in the future we could rename "score" to "data" but I don't think it matters

Adimote commented on scoring/score.py:
```diff
@@ -13,21 +25,182 @@ def __init__(self, message: str, *, code: str) -> None:
+        return right
+
+    left = ", ".join(strings[:-1])
+    return f"{left} {separator} {strings[-1]}"
```
[extra-nit] oxford comma? :)
```suggestion
    return f"{left}{',' if len(strings) > 2 else ''} {separator} {strings[-1]}"
```

PeterJCLaw replied:
no thanks :)

Adimote commented on scoring/score.py:
```diff
@@ -13,21 +25,182 @@ def __init__(self, message: str, *, code: str) -> None:
+            )
+
+        # Check that the "highest" pallet in districts which don't have a
+        # high-rises has another pallet to be placed on top of (pallets on the
```
[nit] grammar fix
```suggestion
        # high-rise has another pallet to be placed on top of (pallets on the
```

Adimote commented on scoring/score.py:
```diff
@@ -13,21 +25,182 @@ def __init__(self, message: str, *, code: str) -> None:
+        bad_highest = {}
+        for name, district in self._districts.items():
+            highest = district['highest']
+            if highest and highest not in ZONE_COLOURS:
```
[nit] I think this is also valid?
```suggestion
            if highest not in ZONE_COLOURS:
```

PeterJCLaw replied:
I don't think we want to alert about cases where `highest` is empty string (`''`) -- implying that there was no highest recorded? The validity of not having a highest pallet is checked elsewhere.

Adimote replied:
oop you're right, I think I misread it, never mind!

Adimote commented:
Looks good overall, I checked the scoring logic against the rules pretty thoroughly so I'm happy with this.

WillB97 commented:
Looks like the checks are:
- Every district is included in the scoring form.
- Every specified pallet is a valid colour.
- If a district has a highest token:
    - It is one of the valid colours and there is only one entry for that district.
    - That pallet colour is present in the district.
    - There is something to place the highest token on, either another token or a high-rise.
- There are not more pallets of any colour than the arena contains.

PeterJCLaw committed:
Clarify docstring

PeterJCLaw committed:
Fix typo

Co-authored-by: Andy Barrett-Sprot <Adimote@users.noreply.github.com>

PeterJCLaw committed:
Shadow the pallet count inputs in their colours

This aims to help use of the scorer during the virtual competition,
where we don't have the instructions on the physical sheet guiding
users on which colours are what.
